pkg/cloudprovider/provider/alibaba/provider.go

```diff
@@ -0,0 +1,413 @@
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rawConfig := alibabatypes.RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.AccessKeyID, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.AccessKeyID, "ALIBABA_ACCESS_KEY_ID")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"AccessKeyID\" field, error = %v", err)
+	}
+	c.AccessKeySecret, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.AccessKeySecret, "ALIBABA_ACCESS_KEY_SECRET")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"AccessKeySecret\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"instanceType\" field, error = %v", err)
+	}
+	c.RegionID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.RegionID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"regionID\" field, error = %v", err)
+	}
+	c.VSwitchID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.VSwitchID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"vSwitchID\" field, error = %v", err)
+	}
+	c.ZoneID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ZoneID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"zoneID\" field, error = %v", err)
+	}
+	c.InternetMaxBandwidthOut, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InternetMaxBandwidthOut)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"internetMaxBandwidthOut\" field, error = %v", err)
+	}
+	c.Labels = rawConfig.Labels
+	return &c, &pconfig, err
+}
+
+func getClient(regionID, accessKeyID, accessKeySecret string) (*ecs.Client, error) {
+	client, err := ecs.NewClientWithAccessKey(regionID, accessKeyID, accessKeySecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Alibaba cloud client: %v", err)
+	}
+	return client, nil
+}
+
+func getInstance(client *ecs.Client, instanceName string) (*ecs.Instance, error) {
+	describeInstanceRequest := ecs.CreateDescribeInstancesRequest()
+	describeInstanceRequest.InstanceName = instanceName
+
+	response, err := client.DescribeInstances(describeInstanceRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to describe instance with instanceName: %s: %v", instanceName, err)
+	}
+
+	if response.Instances.Instance == nil || len(response.Instances.Instance) == 0 {
+		return nil, errors.ErrInstanceNotFound
+	}
+
+	return &response.Instances.Instance[0], nil
+}
+
+func getImageIDForOS(os providerconfigtypes.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfigtypes.OperatingSystemUbuntu:
+		return "ubuntu_18_04_64_20G_alibase_20190624.vhd", nil
```

Comment:
This will get outdated quickly, can we do one of:
* Look this up dynamically 
* Allow users to set it
* Allow users to set it and fall back to looking it up dynamicall (e.G. AWS does this)

?